modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/ReadWriteTransactionImpl.java

```diff
@@ -51,6 +53,12 @@ public class ReadWriteTransactionImpl extends IgniteAbstractTransactionImpl {
     /** Reference to the partition that stores the transaction state. */
     private final AtomicReference<ReplicationGroupId> commitPartitionRef = new AtomicReference<>();
 
+    /** The future used on repeated commit/rollback. */
+    private final AtomicReference<CompletableFuture<Void>> finishFut = new AtomicReference<>();
+
+    /** {@code true} if commit is started. */
+    private volatile boolean commitState;
```

Comment:
The same as it is for Client transaction: the variable is not really required.

Reply:
fixed